fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs

```diff
@@ -0,0 +1,226 @@
+                     autoReplenishment: true));
+    }
+    else
+    {
+        return RateLimitPartition.CreateNoLimiter<IPAddress>(IPAddress.Loopback);
+    }
+});
+
+app.UseRateLimiter(options);
+
+app.MapRazorPages().RequireRateLimiting(userPolicyName);
+app.MapDefaultControllerRoute();
+
+static string GetUserEndPoint(HttpContext context) =>
+    $"User {context.User?.Identity?.Name ?? "Anonymous"}  endpoint: {context.Request.Path}" +
+    $" {context.Connection.RemoteIpAddress}";
+
+app.MapGet("/a", (HttpContext context) => $"{GetUserEndPoint(context)}")
+    .RequireRateLimiting(userPolicyName);
+
+app.MapGet("/b", (HttpContext context) => $"{GetUserEndPoint(context)}")
+    .RequireRateLimiting(userPolicyName);
+
+app.MapGet("/c", (HttpContext context) => $"{GetUserEndPoint(context)}");
+
+app.Run();
+// </snippet>
+// </snippet_1>
+#elif ADMIN
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.RateLimiting;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Primitives;
+using System.Threading.RateLimiting;
+using WebRateLimitAuth.Data;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddAuthorization(o => o.AddPolicy("AdminsOnly",
+                                  b => b.RequireClaim("admin", "true")));
+
+var connectionString = builder.Configuration.GetConnectionString("DefaultConnection") 
+    ?? throw new InvalidOperationException("Connection string 'DefaultConnection' not found.");
+builder.Services.AddDbContext<ApplicationDbContext>(options =>
+    options.UseSqlServer(connectionString));
+builder.Services.AddDatabaseDeveloperPageExceptionFilter();
+
+builder.Services.AddDefaultIdentity<IdentityUser>(options => 
+                  options.SignIn.RequireConfirmedAccount = true)
+    .AddEntityFrameworkStores<ApplicationDbContext>();
+
+builder.Services.AddRazorPages();
+builder.Services.AddControllersWithViews();
+
+var app = builder.Build();
+
+if (app.Environment.IsDevelopment())
+{
+    app.UseMigrationsEndPoint();
+}
+else
+{
+    app.UseExceptionHandler("/Error");
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+
+// <snippet_adm>
+app.UseAuthentication();
+app.UseAuthorization();
+
+var getPolicyName = "get";
+var adminPolicyName = "admin";
+var postPolicyName = "post";
+
+app.UseRateLimiter(new RateLimiterOptions()
+    .AddConcurrencyLimiter(policyName: getPolicyName,
+          new ConcurrencyLimiterOptions(permitLimit: 2,
+          queueProcessingOrder: QueueProcessingOrder.OldestFirst,          
+          queueLimit: 2))
+    .AddNoLimiter(policyName: adminPolicyName)
+    .AddPolicy(policyName: postPolicyName, partitioner: httpContext =>
+    {
+        var myToken = httpContext.Request.Headers["my-token"].ToString();
+        if (!StringValues.IsNullOrEmpty(myToken))
+        {
+            return RateLimitPartition.CreateTokenBucketLimiter( myToken, key =>
```

Comment:
Since most users might just copy-paste this code, I'm thinking about the security of this limiter. Should we make sure that user is authenticated (`context.User.Identity.IsAuthenticated`) and then limit by the token? Here we are creating a dictionary based on user input. I'm just want to make sure that this is not an attack vector.
Or maybe we should get token from where it's saved [here](https://github.com/dotnet/aspnetcore/blob/fd1891536f27e959d14a140ff9307b6a21191de9/src/Security/Authentication/JwtBearer/src/JwtBearerHandler.cs#L152-L158) instead of directly from the header. Although I don't know the best way to access `AuthenticationProperties`.

```csharp
HttpContext.Features.Get<IAuthenticateResultFeature>().AuthenticateResult.Properties.GetTokenValue("access_token");
```

Reply:
@Kahbazi please review my update.